### DISCUSSION TITLE
Dart - How to access FSharpResult in Dart

### DISCUSSION THREAD

**Discussion by sandeepc24** [fable-compiler/Fable#3374] [General]
I am looking at reusing some F# code in my Flutter app and there are some F# functions that return FSharpResult, how can I access Ok/Error values in Dart?

**Comment by alfonsogarciacaro** [comment]
This is a good question :)

You could check the `tag` field as compiled code does or use the helpers from fable_library, but probably neither is particularly nice to use from Dart code. We could make some work to improve pattern matching from native code, but Dart 3 is bringing [sealed classes and pattern matching](https://medium.com/dartlang/dart-3-alpha-f1458fb9d232) which will work very similar to F# unions, so it's better to wait and support this when Dart 3 is available.

For now, your best option is probably to declare some type aliases and extensions to wrap fable_library types and functions to make them more palatable when using them from Dart. For example, `Result.map`

```dart
// The location of fable_modules will depend on the outDir of the Fable compilation
// We are also considering to distribute fable_library as a package in pub.dev
import './fable_modules/fable_library/Choice.dart' as choice;

typedef Result<Ok, Error> = choice.FSharpResult$2<Ok, Error>;

extension ResultExt<Ok, Error> on Result<Ok, Error> {
  Result<Ok2, Error> map<Ok2>(Ok2 Function(Ok) f) {
    return choice.Result_Map(f, this);
  }
}


// Usage
Result<double, String> test(Result<int, String> res) {
  return res.map((x) => x + 5);
}
```

**Comment by ReyCarrozzakF** [comment]
FSharpResult is a type used in F# to represent the result of an operation, which can be either successful (represented by an Ok value) or failed (represented by an Error value). To access these values in Dart, you can define a corresponding Result class and use it to represent the FSharpResult type.

**Reply by alfonsogarciacaro** [reply]
Thanks @ReyCarrozzakF! I assume @sandeepc24 means result types returned by F# code. In that case it won't work to define your own class because it will be a different type. That's why you need to reference the type from fable-library, which is a bit cumbersome right now, but hopefully we can make it easier in the future.
